crates/holo_hash/src/hash.rs

```diff
@@ -1,33 +1,85 @@
-use crate::{has_hash::HasHash, HashType, PrimitiveHashType};
-
-pub(crate) const HASH_CORE_LEN: usize = 32;
-pub(crate) const HASH_LOC_LEN: usize = 4;
-
-/// Length of the a full HoloHash bytes
-pub const HOLO_HASH_SERIALIZED_LEN: usize = HASH_CORE_LEN + HASH_LOC_LEN;
+//! Defines the HoloHash type, used for all hashes in Holochain.
+//!
+//! HoloHashes come in a variety of types. See the `hash_type::primitive`
+//! module for the full list.
+//!
+//! HoloHashes are serialized as a plain 39-byte sequence.
+//! The structure is like so:
+//!
+//! ```text
+//! PPPCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCLLLL
+//! ^  ^                                  ^
+//!  \  \---------"untyped"--------------/
+//!   \                                 /
+//!    \-------------"full"------------/
+//!
+//! P: 3 byte prefix to indicate hash type
+//! C: 32 byte hash, the "core"
+//! L: 4 byte hash of the core hash, for DHT location
+//! ```
+//!
+//! The 36 bytes which exclude the initial 3-byte type prefix are known
+//! throughout the codebase as the "untyped" hash
+//!
+//! The complete 39 bytes together are known as the "full" hash
+
+use crate::{error::HoloHashResult, has_hash::HasHash, HashType, PrimitiveHashType};
+
+/// Length of the prefix bytes (3)
+pub const HOLO_HASH_PREFIX_LEN: usize = 3;
+
+/// Length of the core bytes (32)
+pub const HOLO_HASH_CORE_LEN: usize = 32;
+
+/// Length of the location bytes (4)
+pub const HOLO_HASH_LOC_LEN: usize = 4;
+
+/// Length of the core bytes + the loc bytes (36 = 32 + 4),
+/// i.e. everything except the type prefix
+pub const HOLO_HASH_UNTYPED_LEN: usize = HOLO_HASH_CORE_LEN + HOLO_HASH_LOC_LEN; // 36
+
+/// Length of the full HoloHash bytes (39 = 3 + 32 + 4)
+pub const HOLO_HASH_FULL_LEN: usize = HOLO_HASH_PREFIX_LEN + HOLO_HASH_CORE_LEN + HOLO_HASH_LOC_LEN;
 
 /// A HoloHash contains a vector of 36 bytes representing a 32-byte blake2b hash
 /// plus 4 bytes representing a DHT location. It also contains a zero-sized
 /// type which specifies what it is a hash of.
-// TODO: make holochain_serial! / the derive able to deal with a type param
-// or if not, implement the TryFroms manually...
-#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
-pub struct HoloHash<T> {
-    #[serde(with = "serde_bytes")]
+#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct HoloHash<T: HashType> {
```

Comment:
If there's a reason we're not applying Searialize/Deserialize to this level, do we want to add a doc comment about it so someone isn't tempted to re-apply in the future?

Reply:
There's custom serialization/deserialization over in ser.rs, so if anyone tries to derive it they'll get an error about duplicate impls that'll show line numbers for both

Reply:
I can add a comment that just points to the custom impls though :+1: